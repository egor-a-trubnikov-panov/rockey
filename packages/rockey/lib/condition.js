import interpolateString from './utils/interpolateString';

const condition = conditionFunc => (...args) => {
  if (conditionFunc()) {
    return interpolateString(...args);
  } else {
    return null;
  }
};

export default condition;
