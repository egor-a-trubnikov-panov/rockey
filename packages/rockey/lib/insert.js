import { staticRule } from './rule';

export const insert = (strings, ...values) => {
  const css = staticRule(strings, ...values);
  css.getClassList();
};

export default insert;
