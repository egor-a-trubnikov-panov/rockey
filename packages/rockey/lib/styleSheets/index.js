import StyleSheet from './StyleSheet';

const sheet = new StyleSheet();

let insertWithQueue = false;

export const insert = precss => {
  if (insertWithQueue) {
    sheet.insertWithQueue(precss);
  } else {
    sheet.insert(precss);
  }
};

export const getRules = css => sheet.getRules();
export const clearStyles = css => sheet.clear();

export const speedy = () => sheet.enableSpeedy();
export const useQueue = () => {
  insertWithQueue = true;
};

export default insert;
