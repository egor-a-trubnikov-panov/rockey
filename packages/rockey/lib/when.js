import isFunction from 'lodash/isFunction';
import interpolateString from './utils/interpolateString';

const when = (...args) => {
  let displayName = null;
  let mixinFunc = null;

  if (args.length === 1) {
    mixinFunc = args[0];
  } else {
    displayName = args[0];
    mixinFunc = args[1];
  }

  return (strings, ...values) => {
    let inline = null;
    let inlineAsFunction = null;

    if (isFunction(strings)) {
      inlineAsFunction = strings;
    } else {
      inline = interpolateString(strings, ...values);
    }

    const anonWhen = (...props) => {
      if (mixinFunc(...props)) {
        return inlineAsFunction ? inlineAsFunction(...props) : inline;
      }

      return null;
    };

    if (displayName) {
      anonWhen.displayName = displayName;
    } else {
      anonWhen.displayName = `anonWhen`;
    }

    return anonWhen;
  };
};

export default when;
