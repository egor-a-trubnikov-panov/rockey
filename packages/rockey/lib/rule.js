import createParser from 'rockey-css-parse';
import isString from 'lodash/isString';
import * as classnames from 'rockey-css-parse/utils/classnames';

import {
  getClassName,
  getAnimationName,
  getMixinClassName,
} from './css/getClassName';

import insert from './styleSheets';
import wrapTemplateStrings from './utils/wrapTemplateStrings';

import vendorPrefix from './plugins/vendorPrefix';
import validateCSSRule from './plugins/validateCSSRule';

const plugins = [];

plugins.push(vendorPrefix());

// if (process.env.NODE_ENV !== 'production') {
//   plugins.push(validateCSSRule());
// }

const parse = createParser({
  getClassName,
  getMixinClassName,
  getAnimationName,
  plugins,
});

const getSelector = classnames.getSelector(getClassName);

export default function rule(raw, ...values) {
  // export default function rule(...args) {
  let parent = null;
  let cached = null;
  let inserted = false;
  let isAlreadyParsed = false;
  let addedMixins = [];

  if (Array.isArray(raw)) {
    isAlreadyParsed = !isString(raw[0]);
  } else if (isString(raw)) {
  } else {
    throw new Error('rockey-rule: wrong call type');
  }

  const key = JSON.stringify({
    raw,
    values,
  });

  return {
    wrapWith(displayName) {
      if (cached) {
        throw new Error(
          'Can not wrap because rule is already parsed and inserted'
        );
      }

      if (isAlreadyParsed) {
        const selector = getSelector(displayName);

        raw.forEach(precss => {
          precss.selector = precss.selector.map(p => `${selector} ${p}`);
          precss.root = [displayName];

          precss.mixins.forEach(m => {
            m.updateSelector(ms => `${selector} ${ms}`);
          });
        });
      } else {
        raw = wrapTemplateStrings(displayName, raw);
      }
    },

    addParent(rule) {
      parent = rule;
    },

    parse() {
      let precss = null;
      let classList = null;

      if (!cached) {
        if (isAlreadyParsed) {
          precss = raw;
          classList = {};

          raw.forEach(part => {
            part.root.forEach(c => {
              if (!classList[c]) {
                classList[c] = getClassName(c);
              }
            });
          });
        } else {
          let parsed = parse(raw, ...values);

          precss = parsed.precss;
          classList = parsed.classList;
        }

        if (addedMixins.length) {
          Object.keys(classList).forEach(displayName => {
            addedMixins.forEach(addedMixin => {
              const css = rule`
                ${displayName} {
                  ${addedMixin}
                }
              `;

              css.parse().precss.forEach(p => {
                precss.push(p);
              });

              // precss = precss.concat(css.parse().precss);
            });
          });
        }

        cached = {
          precss,
          classList,
        };
      } else {
        precss = cached.precss;
        classList = cached.classList;
      }

      return { precss, classList };
    },

    transform(transformFunc) {
      let precss = {};

      if (isAlreadyParsed) {
        precss = raw;
      } else {
        const parsed = this.parse();
        precss = parsed.precss;
      }

      const tree = {};

      precss.forEach(part => {
        part.root.forEach(key => {
          if (!tree[key]) {
            tree[key] = [];
          }
          tree[key].push(part);
        });
      });

      return transformFunc(tree, transformed => {
        return rule(Array.isArray(transformed) ? transformed : [transformed]);
      });
    },

    addMixins(mixins) {
      addedMixins = addedMixins.concat(mixins);
    },

    getClassList(props = {}) {
      let parentClasses = [];

      if (parent) {
        const parentClassList = parent.getClassList(props);
        Object.keys(parentClassList).forEach(name => {
          parentClasses.push(parentClassList[name]);
        });
      }

      const { classList, precss } = this.parse();

      if (!inserted) {
        if (precss.length) {
          insert(precss);
        }

        inserted = true;
      }

      let resultClassList = {};

      if (parent) {
        let parentClassList = parentClasses.join(' ');
        if (parentClassList) {
          Object.keys(classList).forEach(name => {
            resultClassList[name] = `${classList[name]} ${parentClassList}`;
          });
        }
      } else {
        resultClassList = Object.assign({}, classList);
      }

      let mixinsPrecss = [];

      precss.forEach(pre => {
        if (pre.mixins) {
          pre.mixins.forEach(mixin => {
            const result = mixin(props);

            if (result) {
              if (result.className) {
                pre.root.forEach(c => {
                  resultClassList[c] += ` ${result.className}`;
                });
              }

              if (result.precss) {
                result.precss.forEach(p => {
                  mixinsPrecss.push(p);
                });
              }
            }
          });
        }
      });

      if (mixinsPrecss.length) {
        insert(mixinsPrecss);
      }

      return resultClassList;
    },
  };
}
