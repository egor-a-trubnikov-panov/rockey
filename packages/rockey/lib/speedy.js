import { speedy } from './styleSheets';
speedy();
