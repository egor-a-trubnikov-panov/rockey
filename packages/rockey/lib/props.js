import when from './when';

const props = name => {
  return when(name, () => true);
};

export default props;
