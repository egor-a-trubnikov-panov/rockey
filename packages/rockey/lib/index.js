import condition from './condition';
import rule from './rule';
import props from './props';
import insert from './insert';
import when from './when';

const rockey = {
  condition,
  rule,
  insert,
  when,
};

export default rockey;
export { condition };
export { insert };
export { when };
export { rule };
export { props };
