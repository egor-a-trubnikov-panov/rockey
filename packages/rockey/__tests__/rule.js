import rule, { staticRule, clearStylesCache } from '../lib/rule';
import when from '../lib/when';
import * as styleSheetsModule from '../lib/styleSheets';

jest.mock('../lib/utils/hash', () => {
  return () => '{{ hash }}';
});

const insertRules = jest.fn();
const insertMixins = jest.fn();

styleSheetsModule.insertRules = insertRules;
styleSheetsModule.insertMixins = insertMixins;

console.warn = jest.fn();

const primary = props => props.isPrimary ? `color: blue` : null;

describe('rule', () => {
  beforeEach(() => {
    insertRules.mockClear();
    insertMixins.mockClear();
    clearStylesCache();
  });

  it('empty rule', () => {
    const css = rule``;
    const classList = css.getClassList();

    expect(classList).toEqual({});
  });

  it('empty rule with wrap', () => {
    const css = rule``;
    css.wrapWith('Layer');

    const classList = css.getClassList();

    expect(classList).toEqual({
      Layer: ['Layer-{{ hash }}'],
    });

    expect(insertRules.mock.calls.length).toEqual(0);
  });

  it('empty rule with parent', () => {
    const css = rule``;
    css.addParent(
      rule`
      Layer {
        color: green
      }
    `
    );

    const classList = css.getClassList();

    expect(classList).toEqual({});

    expect(insertRules.mock.calls.length).toEqual(1);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Layer-{{ hash }}': {
        color: 'green',
      },
    });
  });

  // ----

  it('classList without mixins', () => {
    const css = rule`
      Button {
        color: green;
      }

      Layer {
        color: red;
      }
    `;

    const classList = css.getClassList();

    expect(classList).toEqual({
      Button: ['Button-{{ hash }}'],
      Layer: ['Layer-{{ hash }}'],
    });

    expect(insertRules.mock.calls.length).toEqual(1);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Button-{{ hash }}': {
        color: 'green',
      },
      '.Layer-{{ hash }}': {
        color: 'red',
      },
    });
  });

  it('classList with mixins #1', () => {
    const css = rule`
      Button {
        color: green;

        ${primary}
      }

      Layer {
        color: red;

        ${primary}
      }
    `;

    const classList = css.getClassList();

    expect(classList).toEqual({
      Button: ['Button-{{ hash }}'],
      Layer: ['Layer-{{ hash }}'],
    });

    expect(insertRules.mock.calls.length).toEqual(1);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Button-{{ hash }}': {
        color: 'green',
      },
      '.Layer-{{ hash }}': {
        color: 'red',
      },
    });
  });

  it('classList with proced mixins #1', () => {
    const css = rule`
      Button {
        color: green;

        ${primary}
      }

      Layer {
        color: red;

        ${primary}
      }
    `;

    const classList = css.getClassList({
      isPrimary: true,
    });

    expect(classList).toEqual({
      Button: ['Button-{{ hash }}', 'Mixin-primary-{{ hash }}-1'],
      Layer: ['Layer-{{ hash }}', 'Mixin-primary-{{ hash }}-1'],
    });

    expect(insertRules.mock.calls.length).toEqual(1);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Button-{{ hash }}': {
        color: 'green',
      },
      '.Layer-{{ hash }}': {
        color: 'red',
      },
    });

    expect(insertMixins.mock.calls.length).toEqual(1);
    expect(insertMixins.mock.calls[0][0]).toEqual({
      '.Mixin-primary-{{ hash }}-1.Button-{{ hash }}': {
        color: 'blue',
      },
      '.Mixin-primary-{{ hash }}-1.Layer-{{ hash }}': {
        color: 'blue',
      },
    });
  });

  it('classList with mixins #2', () => {
    const css = rule`
      Button {
        color: green;

        ${primary}
        ${primary}
        ${props => props.isRaised ? `color: green` : null}
      }

      Layer {
        color: red;

        ${primary}
        ${primary}
        ${props => props.isRaised ? `color: green` : null}
      }
    `;

    const classList = css.getClassList();

    expect(classList).toEqual({
      Button: ['Button-{{ hash }}'],
      Layer: ['Layer-{{ hash }}'],
    });

    expect(insertRules.mock.calls.length).toEqual(1);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Button-{{ hash }}': {
        color: 'green',
      },
      '.Layer-{{ hash }}': {
        color: 'red',
      },
    });
  });

  it('classList with proced mixins #2', () => {
    const css = rule`
      Button {
        color: green;

        ${primary}
        ${primary}
        ${props => props.isRaised ? `color: green` : null}
      }

      Layer {
        color: red;

        ${primary}
        ${primary}
        ${props => props.isRaised ? `color: green` : null}
      }
    `;

    const classList = css.getClassList({
      isPrimary: true,
      isRaised: true,
    });

    expect(classList).toEqual({
      Button: [
        'Button-{{ hash }}',
        'Mixin-primary-{{ hash }}-1',
        'Mixin-primary-{{ hash }}-1',
        'Mixin-anon-{{ hash }}-1',
      ],
      Layer: [
        'Layer-{{ hash }}',
        'Mixin-primary-{{ hash }}-1',
        'Mixin-primary-{{ hash }}-1',
        'Mixin-anon-{{ hash }}-1',
      ],
    });

    expect(insertRules.mock.calls.length).toEqual(1);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Button-{{ hash }}': {
        color: 'green',
      },
      '.Layer-{{ hash }}': {
        color: 'red',
      },
    });
  });

  // ----

  it('classList with anon when', () => {
    const color = 'blue';

    const css = rule`
      Button {
        color: red;

        ${when(props => props.isPrimary)`
            color: ${color};
          `}
      }
    `;

    expect(css.getClassList()).toEqual({
      Button: ['Button-{{ hash }}'],
    });
  });

  it('classList with proced anon when', () => {
    const color = 'blue';

    const css = rule`
      Button {
        color: red;

        ${when(props => props.isPrimary)`
            color: ${color};
          `}
      }
    `;

    expect(
      css.getClassList({
        isPrimary: true,
      })
    ).toEqual({
      Button: ['Button-{{ hash }}', 'Mixin-anonWhen-{{ hash }}-1'],
    });
  });

  it('classList with named when', () => {
    const color = 'blue';

    const css = rule`
      Button {
        color: red;

        ${when('isPrimary', props => props.isPrimary)`
          color: ${color};
          `}
      }
    `;

    expect(css.getClassList()).toEqual({
      Button: ['Button-{{ hash }}'],
    });
  });

  it('classList with proced named when', () => {
    const color = 'blue';

    const css = rule`
      Button {
        color: red;

        ${when('isPrimary', props => props.isPrimary)`
          color: ${color};
          `}
      }
    `;

    expect(
      css.getClassList({
        isPrimary: true,
      })
    ).toEqual({
      Button: ['Button-{{ hash }}', 'Mixin-isPrimary-{{ hash }}-1'],
    });
  });

  it('classList with proced mixin', () => {
    const background = '#ccc';

    const css = rule`
      Button {
        color: red;

        ${primary}

        ${function isRaised(props) {
      return props.isRaised ? `background: ${background}` : null;
    }}
      }
    `;

    expect(
      css.getClassList({
        isPrimary: true,
        isRaised: true,
      })
    ).toEqual({
      Button: [
        'Button-{{ hash }}',
        'Mixin-primary-{{ hash }}-1',
        'Mixin-isRaised-{{ hash }}-1',
      ],
    });

    expect(insertRules.mock.calls.length).toEqual(1);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Button-{{ hash }}': {
        color: 'red',
      },
    });

    expect(insertMixins.mock.calls.length).toEqual(1);
    expect(insertMixins.mock.calls[0][0]).toEqual({
      '.Mixin-primary-{{ hash }}-1.Button-{{ hash }}': {
        color: 'blue',
      },
      '.Mixin-isRaised-{{ hash }}-1.Button-{{ hash }}': {
        background: '#ccc',
      },
    });
  });

  it('classList with proced mixin and wrapped with component', () => {
    const css = rule`
      Button {
        color: red;

        ${primary}
      }
    `;

    css.wrapWith('Layer');

    expect(
      css.getClassList({
        isPrimary: true,
      })
    ).toEqual({
      Layer: ['Layer-{{ hash }}', 'Mixin-primary-{{ hash }}-1'],
    });

    expect(insertRules.mock.calls.length).toEqual(1);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Layer-{{ hash }} .Button-{{ hash }}': {
        color: 'red',
      },
    });

    expect(insertMixins.mock.calls.length).toEqual(1);
    expect(insertMixins.mock.calls[0][0]).toEqual({
      '.Mixin-primary-{{ hash }}-1.Layer-{{ hash }} .Button-{{ hash }}': {
        color: 'blue',
      },
    });
  });

  it('staticRule', () => {
    const css = staticRule`
      Button {
        color: red;
      }
    `;

    expect(css.getClassList()).toEqual({
      Button: ['Button-{{ hash }}'],
    });

    expect(insertRules.mock.calls.length).toEqual(1);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Button-{{ hash }}': {
        color: 'red',
      },
    });
  });

  it('staticRule with mixin', () => {
    expect(
      () => staticRule`
      Button {
        color: red;
      }

      ${primary}
    `
    ).toThrow('Static rule should not contain mixins');
  });

  it('staticRule with variables', () => {
    const color = 'green';
    const css = staticRule`
      Button {
        color: ${color};
      }
    `;

    expect(css.getClassList()).toEqual({
      Button: ['Button-{{ hash }}'],
    });

    expect(insertRules.mock.calls.length).toEqual(1);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Button-{{ hash }}': {
        color: 'green',
      },
    });
  });

  it('rule with dynamic mixins', () => {
    const css = rule`
      Button {
        color: red;

        ${function colored(props) {
      return `color: ${props.color}`;
    }}
      }
    `;

    expect(
      css.getClassList({
        color: '#ffcc33',
      })
    ).toEqual({
      Button: ['Button-{{ hash }}', 'Mixin-colored-{{ hash }}-1'],
    });

    expect(
      css.getClassList({
        color: '#cc0000',
      })
    ).toEqual({
      Button: ['Button-{{ hash }}', 'Mixin-colored-{{ hash }}-2'],
    });

    expect(
      css.getClassList({
        color: '#3399ff',
      })
    ).toEqual({
      Button: ['Button-{{ hash }}', 'Mixin-colored-{{ hash }}-3'],
    });

    expect(
      css.getClassList({
        color: '#ffcc33',
      })
    ).toEqual({
      Button: ['Button-{{ hash }}', 'Mixin-colored-{{ hash }}-1'],
    });

    expect(insertRules.mock.calls.length).toEqual(1);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Button-{{ hash }}': {
        color: 'red',
      },
    });

    expect(insertMixins.mock.calls.length).toEqual(3);
    expect(insertMixins.mock.calls[0][0]).toEqual({
      '.Mixin-colored-{{ hash }}-1.Button-{{ hash }}': {
        color: '#ffcc33',
      },
    });
    expect(insertMixins.mock.calls[1][0]).toEqual({
      '.Mixin-colored-{{ hash }}-2.Button-{{ hash }}': {
        color: '#cc0000',
      },
    });
    expect(insertMixins.mock.calls[2][0]).toEqual({
      '.Mixin-colored-{{ hash }}-3.Button-{{ hash }}': {
        color: '#3399ff',
      },
    });
  });

  it('rule add parent', () => {
    const parent = rule`
      Button {
        color: red;
        font-size: 10px;

        ${primary}
      }
    `;

    const child = rule`
      PrimaryButton {
        color: blue;
      }
    `;

    child.addParent(parent);

    expect(child.getClassList()).toEqual({
      PrimaryButton: ['PrimaryButton-{{ hash }}', 'Button-{{ hash }}'],
    });

    expect(insertRules.mock.calls.length).toEqual(2);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Button-{{ hash }}': {
        color: 'red',
        'font-size': '10px',
      },
    });

    expect(insertRules.mock.calls[1][0]).toEqual({
      '.PrimaryButton-{{ hash }}': {
        color: 'blue',
      },
    });

    expect(insertMixins.mock.calls.length).toEqual(0);
  });

  it('rule add parent with proced mixin', () => {
    const parent = rule`
      Button {
        color: red;
        font-size: 10px;

        ${function colored(props) {
      return `color: ${props.color}`;
    }}
      }
    `;

    const child = rule`
      PrimaryButton {
        color: blue;
      }
    `;

    child.addParent(parent);

    expect(
      child.getClassList({
        color: '#ffcc33',
      })
    ).toEqual({
      PrimaryButton: [
        'PrimaryButton-{{ hash }}',
        'Button-{{ hash }}',
        'Mixin-colored-{{ hash }}-1',
      ],
    });

    expect(insertRules.mock.calls.length).toEqual(2);
    expect(insertRules.mock.calls[0][0]).toEqual({
      '.Button-{{ hash }}': {
        color: 'red',
        'font-size': '10px',
      },
    });

    expect(insertRules.mock.calls[1][0]).toEqual({
      '.PrimaryButton-{{ hash }}': {
        color: 'blue',
      },
    });

    expect(insertMixins.mock.calls.length).toEqual(1);
    expect(insertMixins.mock.calls[0][0]).toEqual({
      '.Mixin-colored-{{ hash }}-1.Button-{{ hash }}': {
        color: '#ffcc33',
      },
    });
  });

  it('mixin inside mixin', () => {
    const css = rule`
      Button {
        color: red;

        ${props => {
      return `
            color: green;
            ${function(props) {
        return 'color: red';
      }}
          `;
    }}
      }
    `;

    expect(() => css.getClassList()).toThrow(
      'Mixin restuls should not contain other mixins'
    );
  });
});
