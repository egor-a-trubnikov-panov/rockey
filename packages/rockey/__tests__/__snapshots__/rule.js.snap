// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rule 1`] = `
Object {
  "Button": "Button-hash m-anon-hash-1 m-background-hash-1 m-anon-hash-1",
}
`;

exports[`rule 2`] = `
Object {
  "Button": "Button-hash m-anon-hash-2 m-background-hash-2",
}
`;

exports[`transform 1`] = `
Object {
  "Button4": "Button4-hash",
}
`;

exports[`transform 2`] = `
Object {
  "Button4": "Button4-hash",
}
`;

exports[`transform 3`] = `
Object {
  "Button3": "Button3-hash",
}
`;

exports[`transform 4`] = `
Object {
  "Button3": "Button3-hash",
}
`;
